Breakout morpheus_llm

AnuradhaKaruppiah opened the pull request:
This PR breaks out the LLM source code into -
1. A separate subdirectory - python/morpheus_llm
2. Compiled as a separate shared lib -
   libmorpheus_llm.so
   |_ llm/__init__.cpython-310-x86_64-linux-gnu.so
3. And setup as a separate python package 
   a. morpheus_llm - all llm modules are now in this namespace
   b. with the following distribution - 
      ./build/python/morpheus_llm/dist/morpheus_llm-24.10.0a0+46.g125a6be9.dirty-py3-none-any.whl

To build and install all components in a dev env you can use these steps (steps also updated in contributing.md) -
1. ./scripts/compile_all.sh
2. pip install -e python/morpheus
3. pip install -e python/morpheus_llm

Items of note -
To ensure right order of build-n-install morpheus_llm has dependencies on the following targets -
1. cudf_helpers
2. morpheus
3. all "py_morpheus" targets
Ref: python/morpheus_llm/morpheus_llm/_lib/cmake/libmorpheus_llm.cmake

Pending (major) items -
1. Need to breakout llm related build/conda environment requirements from the base morpheus requirements
2. Need to breakout llm python tests (cpp tests have been moved as a part of this commit).

AnuradhaKaruppiah commented:
/ok to test

AnuradhaKaruppiah commented:
/ok to test

dagardner-nv reviewed (changes requested):
Looks good, just need to fix the documentation builds.

AnuradhaKaruppiah commented on ci/scripts/github/docs.sh:
```diff
@@ -30,6 +30,7 @@ download_artifact "wheel.tar.bz"
 tar xf "${WORKSPACE_TMP}/wheel.tar.bz"
 
 pip install ${MORPHEUS_ROOT}/${BUILD_DIR}/python/morpheus/dist/*.whl
+pip install ${MORPHEUS_ROOT}/${BUILD_DIR}/python/morpheus_llm/dist/*.whl
```
Ack

AnuradhaKaruppiah replied:
Addressed in [this](https://github.com/nv-morpheus/Morpheus/pull/1853/commits/d7bc5cbefd1a87aedffd4d7a379418e75a42d994) commit

AnuradhaKaruppiah commented on python/CMakeLists.txt:
```diff
@@ -16,5 +16,6 @@
 list(APPEND CMAKE_MESSAGE_CONTEXT "python")
 
 add_subdirectory(morpheus)
+add_subdirectory(morpheus_llm)
```
Ack. I will take care of it in this PR itself. Thx for the review @dagardner-nv

AnuradhaKaruppiah commented:
/ok to test

AnuradhaKaruppiah commented:
/ok to test

AnuradhaKaruppiah commented:
@dagardner-nv Can you please take another look at the review. I have updated it -
1. To include a build flag for morpheus_llm - commits [1](https://github.com/nv-morpheus/Morpheus/pull/1853/commits/0ca14fa0331d7b77635a89176c2de39668f97d56), [2](https://github.com/nv-morpheus/Morpheus/pull/1853/commits/e6ec1dd03d94bc2ad44729be550f8eef4da426d2), [3](https://github.com/nv-morpheus/Morpheus/pull/1853/commits/7ab5797256279485303d3beb96aca465ea8f777f)

2. Changed the include dir name to morpheus_llm to prevent namespace conflict when `MORPHEUS_PYTHON_INPLACE_BUILD=OFF` - commits [1](https://github.com/nv-morpheus/Morpheus/pull/1853/commits/d0fa19cec4544ed84292b5a538f0035ca1376133), [2](https://github.com/nv-morpheus/Morpheus/pull/1853/commits/c6f9aa0138c6dd9564dbd34d1059364b7c97ae85), [3](https://github.com/nv-morpheus/Morpheus/pull/1853/commits/c6f9aa0138c6dd9564dbd34d1059364b7c97ae85)

3. Update the docs build for including morpheus_llm in the py_api.html and python/morpheus_llm/morpheus_llm/_lib while generating the C++ APIs - [commit](https://github.com/nv-morpheus/Morpheus/pull/1853/commits/d7bc5cbefd1a87aedffd4d7a379418e75a42d994)

dagardner-nv commented on docs/CMakeLists.txt:
```diff
@@ -30,15 +30,15 @@ add_custom_target(${PROJECT_NAME}_docs
    COMMENT "Generating documentation with Sphinx"
-   DEPENDS morpheus-package-outputs
+   DEPENDS morpheus-package-outputs morpheus_llm-package-outputs
 )
 
 add_custom_target(${PROJECT_NAME}_docs_linkcheck
    COMMAND
       BUILD_DIR=${CMAKE_CURRENT_BINARY_DIR} ${SPHINX_EXECUTABLE} ${SPHINX_LINKCHECK_ARGS} ${SPHINX_SOURCE} ${SPHINX_LINKCHECK_OUT}
    WORKING_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}
    COMMENT "Checking documentation links with Sphinx"
-   DEPENDS morpheus-package-outputs
+   DEPENDS morpheus-package-outputs morpheus_llm-package-outputs
```
The problem is that `MORPHEUS_BUILD_MORPHEUS_LLM` defaults to `OFF`, if you build with `MORPHEUS_BUILD_MORPHEUS_LLM=OFF` then you get an error:
```
ninja: error: 'docs/morpheus_llm-package-outputs', needed by 'docs/CMakeFiles/morpheus_docs', missing and no known rule to make it
```

AnuradhaKaruppiah replied:
good catch. I missed that. I will make the llm dependency conditional.

dagardner-nv reviewed (approved):
LGTM, just one issue noted above re documentation builds.

AnuradhaKaruppiah commented:
/ok to test

AnuradhaKaruppiah commented:
/ok to test

AnuradhaKaruppiah commented:
/ok to test

AnuradhaKaruppiah commented:
/merge
